lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,472 @@
+class Substrate(Enum):
+    """Determines which type of charm this library is running in.
+
+    Some library functions need to interact with the substrate by reading and writing files. This
+    enum allows the library to make the correct decision on which methods to use depending on the
+    substrate.
+    """
+
+    KUBERNETES = "kubernetes"
+    MACHINE = "machine"
+
+
+class File(Enum):
+    """This enum determines which files are expected of the library to read."""
+
+    CERT = auto()
+    KEY = auto()
+    CA = auto()
+    CSR = auto()
+
+
+class VaultTLSManager(Object):
+    """This class configures the certificates within Vault."""
+
+    def __init__(self, charm, peer_relation: str, substrate: Substrate):
+        """Manager of MongoDB TLS relation and configuration."""
+        super().__init__(charm, "tls")
+        self.charm = charm
+        self.substrate = substrate
+        self.peer_relation = peer_relation
+        self.tls_access = TLSCertificatesRequiresV2(charm, TLS_CERTIFICATE_ACCESS_RELATION_NAME)
+        self.certificate_transfer = CertificateTransferProvides(charm, SEND_CA_CERT_RELATION_NAME)
+
+        self.framework.observe(
+            self.charm.on[TLS_CERTIFICATE_ACCESS_RELATION_NAME].relation_joined,
+            self._on_certificate_config_changed,
+        )
+        self.framework.observe(
+            self.charm.on[TLS_CERTIFICATE_ACCESS_RELATION_NAME].relation_broken,
+            self._on_tls_access_certificate_relation_broken,
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_available, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_expiring, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.tls_access.on.certificate_invalidated, self._on_certificate_config_changed
+        )
+        self.framework.observe(
+            self.charm.on[SEND_CA_CERT_RELATION_NAME].relation_joined,
+            self._on_certificate_config_changed,
+        )
+
+    def _on_certificate_config_changed(self, event: EventBase):
+        """Handle TLS configuration changes. Makes the charm reconfigure its environment."""
+        self.charm.on.config_changed.emit()
+
+    def _on_tls_access_certificate_relation_broken(self, event: RelationBrokenEvent):
+        """Handles leaving the tls access relation.
+
+        Regenerates self signed certificates from the saved self generated root CA
+        and reloads vault.
+        """
+        if not self.charm._ingress_address:
+            self._remove_all_certs_from_workload()
+            return
+        self._generate_self_signed_certs(self.charm._ingress_address)
+        logger.info(
+            "TLS: Saved Vault generated CA and self signed certificate to %s.",
+            self.charm.unit.name,
+        )
+        self._reload_vault_container()
+
+    def configure_certificates(self, ingress_address: str) -> None:
```

Comment:
The purpose of having this as an argument of the function is to make sure that it can never be set to none when this function is running. This allows any future user to run this function from anywhere and be guaranteed that it will do the correct job. It also lets mypy catch any time a variable is passed here as `Optional[str]` instead of just str, so if this variable was accidentally not set, we will catch it during the `tox static` phase instead of at runtime.